crates/parcel/src/requests/target_request.rs

```diff
@@ -0,0 +1,1107 @@
+  dist: Option<PathBuf>,
+  extensions: Vec<&'a str>,
+  name: &'a str,
+}
+
+struct CustomTarget<'a> {
+  descriptor: &'a TargetDescriptor,
+  name: &'a str,
+}
+
+impl TargetRequest {
+  fn builtin_target_descriptor(&self) -> TargetDescriptor {
+    TargetDescriptor {
+      include_node_modules: Some(IncludeNodeModules::Bool(false)),
+      is_library: Some(true),
+      scope_hoist: Some(true),
+      ..TargetDescriptor::default()
+    }
+  }
+
+  fn builtin_browser_target(
+    &self,
+    descriptor: Option<BuiltInTargetDescriptor>,
+    dist: Option<BrowserField>,
+    name: Option<String>,
+  ) -> BuiltInTarget {
+    BuiltInTarget {
+      descriptor: descriptor.unwrap_or_else(|| {
+        BuiltInTargetDescriptor::TargetDescriptor(TargetDescriptor {
+          context: Some(EnvironmentContext::Browser),
+          ..self.builtin_target_descriptor()
+        })
+      }),
+      dist: dist.and_then(|browser| match browser {
+        BrowserField::EntryPoint(entrypoint) => Some(entrypoint.clone()),
+        BrowserField::ReplacementBySpecifier(replacements) => {
+          name.and_then(|name| replacements.get(&name).map(|v| v.into()))
+        }
+      }),
+      extensions: vec!["cjs", "js", "mjs"],
+      name: "browser",
+    }
+  }
+
+  fn builtin_main_target(
+    &self,
+    descriptor: Option<BuiltInTargetDescriptor>,
+    dist: Option<PathBuf>,
+  ) -> BuiltInTarget {
+    BuiltInTarget {
+      descriptor: descriptor.unwrap_or_else(|| {
+        BuiltInTargetDescriptor::TargetDescriptor(TargetDescriptor {
+          context: Some(EnvironmentContext::Node),
+          ..self.builtin_target_descriptor()
+        })
+      }),
+      dist,
+      extensions: vec!["cjs", "js", "mjs"],
+      name: "main",
+    }
+  }
+
+  fn builtin_module_target(
+    &self,
+    descriptor: Option<BuiltInTargetDescriptor>,
+    dist: Option<PathBuf>,
+  ) -> BuiltInTarget {
+    BuiltInTarget {
+      descriptor: descriptor.unwrap_or_else(|| {
+        BuiltInTargetDescriptor::TargetDescriptor(TargetDescriptor {
+          context: Some(EnvironmentContext::Node),
+          ..self.builtin_target_descriptor()
+        })
+      }),
+      dist,
+      extensions: vec!["js", "mjs"],
+      name: "module",
+    }
+  }
+
+  fn builtin_types_target(
+    &self,
+    descriptor: Option<BuiltInTargetDescriptor>,
+    dist: Option<PathBuf>,
+  ) -> BuiltInTarget {
+    BuiltInTarget {
+      descriptor: descriptor.unwrap_or_else(|| {
+        BuiltInTargetDescriptor::TargetDescriptor(TargetDescriptor {
+          context: Some(EnvironmentContext::Node),
+          ..self.builtin_target_descriptor()
+        })
+      }),
+      dist,
+      extensions: vec!["ts"],
+      name: "types",
+    }
+  }
+
+  fn default_dist_dir(&self, package_path: &Path) -> PathBuf {
+    package_path
+      .parent()
+      .unwrap_or_else(|| &package_path)
+      .join("dist")
+  }
+
+  fn fallback_output_format(&self, context: EnvironmentContext) -> OutputFormat {
```

Comment:
we should prefer free functions for things that don't take `self`

Reply:
imo this is personal preference, but can change it 🤷

Reply:
More or less ; it is a bit of a style preference, but there are many advantages to having free functions - https://www.youtube.com/watch?v=WLDT1lDOsb4

* On lots of languages, non-free function denies some optimisations (not sure if in rust)
* To test the function you do not need to create self
* ...